EKF/control.cpp

```diff
@@ -757,10 +757,19 @@ void Ekf::controlHeightFusion()
 		// we have just switched to using range finder, calculate height sensor offset such that current
 		// measurment matches our current height estimate
 		if (_control_status_prev.flags.rng_hgt != _control_status.flags.rng_hgt) {
-			if (_terrain_initialised) {
-				_hgt_sensor_offset = _terrain_vpos;
+			// use the parameter rng_gnd_clearance if on ground to avoid a noisy offset initialization (e.g. sonar)
+			if (_control_status.flags.in_air) {
+
+				if (_terrain_initialised) {
+
+					_hgt_sensor_offset = _terrain_vpos;
+				} else {
+
+					_hgt_sensor_offset = _R_rng_to_earth_2_2 * _range_sample_delayed.rng + _state.pos(2);
```

Comment:
To me this is not a problem as it is said in the doc that this setting should be used only over flat terrain and that NED origin may move up and down with terrain. So the Z position command becomes a height-above-ground command.

In your example, I agree that a sudden change of altitude is not desired, but 
1. changing the primary height source in flight sounds like a bad idea anyway
2. if not with sudden altitude change, the drone should converge to the commanded height. In your example the drone will keep an altitude error. Let's say the drone is flying at 2m altitude using baro. Then it moves at constant altitude above a 1.5 meter hill, now baro altitude is still 2m, but the sonar returns a height of 0.5m. By switching to range sensor and updating `_hgt_sensor_offset`, the drone will stay 0.5m above ground, which is not the commanded height of 2m.

When sonar becomes primary height sensor, I would think the *baro* offset needs to be updated, and not the sonar offset.